### PR TITLE
feat(endpoints): add Anthropic Messages API (/v1/messages) endpoint

### PR DESCRIPTION
**Pull request by ajcasagrande**
Adds a full-featured **Anthropic Messages API (`/v1/messages`) endpoint** (`--endpoint-type messages`) for benchmarking Anthropic-compatible servers — the real Anthropic API, Dynamo's `/v1/messages`, and other Messages-compatible implementations.

This PR is the **unification of the two independent implementations**: it takes #1054's `MessagesEndpoint` (@hnt2601's commit is this branch's first commit, authorship preserved) as the base, ports in the stronger pieces from the original version of this PR, and fixes everything surfaced in the full review — findings, runtime repros, and a head-to-head comparison of both implementations are posted on [#1054](https://github.com/ai-dynamo/aiperf/pull/1054#issuecomment-4897982561).

```bash
aiperf profile \
    --model claude-sonnet-4-20250514 \
    --endpoint-type messages \
    --streaming \
    --url api.anthropic.com \
    --api-key $ANTHROPIC_API_KEY \
    --request-count 20 \
    --concurrency 4
```

## Endpoint (`src/aiperf/endpoints/anthropic_messages.py`)

- **Streaming + non-streaming** parsing of the full Messages event vocabulary (typed event/block/delta enums; explicit `error`-event logging; unknown events debug-logged).
- **Auth that works against the real API**: `--api-key` → `x-api-key`, `anthropic-version: 2023-06-01` sent by default (user `--header` overrides win). Harmless for compat servers that require neither.
- **Tool use counts toward client-side OSL in both modes**: streaming `input_json_delta` fragments and non-streaming `tool_use` blocks, with chat-endpoint precedence (`reasoning > text+tool > tool > text`) so mixed prose+tool responses count both portions.
- **Extended thinking**: `thinking_delta` → reasoning metrics; `signature` preserved.
- **Split streaming usage merged**: `extract_response_data` folds `message_start` usage keys into the final usage-bearing response, so servers following the docs-canonical shape (output-only `message_delta` usage) don't lose `input_tokens` to the record layer's last-usage-wins merge; cumulative servers (real Anthropic, Dynamo main) are untouched.
- **Images**: data-URI/base64 sources (with parsed `media_type`) and URL sources.
- **System prompts**: top-level `system` string, or per-turn `raw_system` block list (new `Turn` field) for `cache_control`-style per-block extensions.
- **Multi-turn / FORK replay**: `build_assistant_turn` reassembles thinking (+signature) and `tool_use` blocks into `raw_messages`, so a follow-up turn's `tool_result` references aren't orphaned.
- **Payload-input extraction** walks top-level `system`, tool `input_schema`, and `tool_use`/`tool_result` history blocks (on a real agentic-traffic sample the base walk missed ~11.5MB of schema and ~140MB of tool-replay text across 374 requests).
- Per-turn payload contract followed: `model`/`max_tokens`/`extra_body` are dispatch-turn-only; `raw_tools`/`raw_system` walk history. `max_tokens` always sent (required by the API and by Dynamo). `stream` key emitted only when streaming, matching real client wire shape.

## Usage layer (`src/aiperf/common/models/usage_models.py`)

Anthropic and Bedrock report **disjoint** input accounting — `input_tokens` is only the *uncached remainder*; the true prompt is `input + cache_read + cache_creation`. `Usage.prompt_tokens` now re-totalizes those shapes (gated on the vendor cache key names, so subset-semantics vendors are never double-counted), and the raw remainder stays available as `prompt_uncached_tokens`. Without this, on a real cached-traffic sample `usage_prompt_tokens` reported **0.04%** of the true prompt and the overall cache-read percentage computed **227,581%** (correct: 98.2%).

## Mock server

`/v1/messages` route with wire shapes validated against real-world Anthropic Messages traffic: full-key usage in `message_start`, cumulative usage in `message_delta` by default (`--anthropic-split-usage` reverts to the docs-canonical output-only shape for usage-merge regression coverage), Anthropic `stop_reason` vocabulary, required `max_tokens` + `anthropic-version` contract enforcement, disjoint prompt-cache simulation (opt-in via `cache_control`; write-then-read ladders across turns), thinking + signature deltas, chunked `input_json_delta` tool streaming, and synthesized `tool_use` when the request declares tools.

## Docs

- New tutorial: `docs/tutorials/anthropic-messages-endpoint.md` (indexed in `docs/index.yml` + README tutorial index)
- `docs/reference/vendor-usage-fields.md`: disjoint-accounting + native-endpoint notes
- Regenerated `docs/cli-options.md`

## Validation

- 65 endpoint unit tests + 10 mock-route wire-contract tests + 227-case disjoint-usage suite; full unit suite green.
- Parse/format logic validated against a ~950-pair real-world `/v1/messages` request/response replay spanning multiple API versions: zero parse errors, zero unknown events/blocks; 120/120 request bodies reproduce exactly through `format_payload` via `raw_messages`/`raw_tools`/`extra_body`.
- The two behavioral fixes (split-usage merge, mixed text+tool OSL) reproduced end-to-end with the real `aiperf` CLI against the mock server before/after.

## Related

- Unifies and supersedes #1054 — full review findings and implementation comparison in [this comment](https://github.com/ai-dynamo/aiperf/pull/1054#issuecomment-4897982561).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added support for benchmarking Anthropic Messages API endpoints, including streaming and non-streaming requests, tool use, image inputs, and shared system prompts.
  * Added cache-aware usage handling and Anthropic-style response parsing for more accurate metrics.

* **Documentation**
  * Expanded tutorials and reference docs with setup, request formats, streaming events, token accounting, and CLI options for the new endpoint.

* **Tests**
  * Added mock-server and unit coverage for request validation, streaming behavior, tool-use chunks, cache accounting, and usage merging.

### CODE REVIEW

File: tests/aiperf_mock_server/tokens.py [hunk 1]
```diff
@@ -260,6 +261,9 @@ def _extract_request_content(request: RequestT) -> tuple[str, int | None]:
         return text, None
     elif isinstance(request, ImageGenerationRequest):
         return request.prompt, None
+    elif isinstance(request, AnthropicMessagesRequest):
+        text = _extract_chat_messages(request.messages)
+        return text, request.max_output_tokens
```
[the remark this answers is not in the document]
Comment (ajcasagrande) [reply tests/aiperf_mock_server/tokens.py:378]: Confirmed valid — the thinking/signature emit paths in the route were unreachable for Anthropic requests since `_generate_reasoning_tokens` only fired for `ChatCompletionRequest`. Fixed in 4015bc48c: thinking is now opt-in per request via `thinking.budget_tokens` (the API's actual trigger — not model-name matching, which is an OpenAI-ism), capped at `max_tokens - 1` so the text block always retains budget, mirroring the API's `budget_tokens < max_tokens` invariant. Added three wire-contract tests: streaming thinking+signature deltas, non-streaming thinking block, and disabled/absent emitting none.

File: tests/unit/endpoints/test_messages_endpoint.py [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+                results.append(parsed)
+
+        # message_start (usage) + 2 text_deltas + message_delta (usage)
+        assert len(results) == 4
+        assert results[0].usage is not None  # message_start
+        assert isinstance(results[1].data, TextResponseData)
+        assert results[1].data.text == "Hello"
+        assert isinstance(results[2].data, TextResponseData)
+        assert results[2].data.text == " world"
+        assert results[3].usage is not None  # message_delta
```
[the remark this answers is not in the document]
Comment (ajcasagrande) [reply tests/unit/endpoints/test_messages_endpoint.py:562]: Added `-> SSEMessage` in 4015bc48c.

File: src/aiperf/endpoints/anthropic_messages.py [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+
+        text = "".join(text_parts) or None
+        thinking = "".join(thinking_parts) or None
+        tool_call_text = "".join(tool_call_parts)
+
+        if thinking:
+            return ReasoningResponseData(content=text, reasoning=thinking)
+        if tool_call_text:
+            return ToolCallResponseData(tool_call_text=tool_call_text, content=text)
+        return self.make_text_response_data(text)
```
[the remark this answers is not in the document]
Comment (ajcasagrande) [reply src/aiperf/endpoints/anthropic_messages.py:743]: Fixed in 4015bc48c — per-block accumulation extracted into a module-level `_accumulate_content_block` helper (module-level to match the other inlined helpers rather than a staticmethod). `make check-ruff-baselined` is clean again: 0 new violations.

File: tests/aiperf_mock_server/app.py [hunk 4]
```diff
@@ -294,6 +298,183 @@ async def _chat_stream_wrapper(
...
+    if choice_type == "none":
+        return None
+    first = tools[0] if isinstance(tools[0], dict) else {}
+    name = first.get("name") or "mock_tool"
+    return {
+        "type": "tool_use",
+        "id": "toolu_mock_1",
+        "name": name,
+        "input": {"arg": "value"},
+    }
```
[the remark this answers is not in the document]
Comment (ajcasagrande) [reply tests/aiperf_mock_server/app.py:418]: Intended behavior — keeping as-is. Absent `tool_choice` defaults to `{"type":"auto"}` on the real API, so absent and explicit `auto` must behave identically; the proposed change would make them diverge (absent synthesizes, explicit auto doesn't). For a benchmarking mock, deterministic responses beat probabilistic realism: when tools are declared the mock always exercises the tool path unless the client explicitly opts out with `{"type":"none"}`. Documented that rationale in the docstring in 4015bc48c.

File: tests/aiperf_mock_server/utils.py [hunk 5]
```diff
@@ -626,3 +629,271 @@ async def stream_tgi_completion(
...
+
+# Prompt-cache simulation state: (model, prefix_hash) entries seen so far.
+# Server-lifetime persistence, like a real prefix cache without TTL expiry.
+_ANTHROPIC_PREFIX_CACHE: set[tuple[str, str]] = set()
+
+
+def reset_anthropic_prefix_cache() -> None:
+    """Clear the simulated prompt-cache (test isolation)."""
+    _ANTHROPIC_PREFIX_CACHE.clear()
+
```
[the remark this answers is not in the document]
Comment (ajcasagrande) [reply tests/aiperf_mock_server/utils.py:664]: Valid limitation — documented in 4015bc48c: the cache is process-local, so cache benchmarks need `--workers 1` (the default). Not adding a hard config-time guard because the simulation is request-triggered (`cache_control` in the request body), not a server flag, so the config layer can't see whether it will be exercised — unlike `--record-requests`, which is a flag and does force `workers=1`.

File: docs/tutorials/anthropic-messages-endpoint.md [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+    ],
+    "messages": [...],
+    "max_tokens": 1024,
+    "stream": false
+}
+```
+
+Notes:
+
+- Latest-non-`None` turn wins — a FORK-mode DAG child that does not re-declare `raw_system` still inherits the parent's value.
```
[the remark this answers is not in the document]
Comment (ajcasagrande) [reply docs/tutorials/anthropic-messages-endpoint.md]: Fixed in 4015bc48c — plus updated helper names in the same tutorial sections that had gone stale when `_anthropic_internals.py` was inlined into the endpoint file (`walk_system` → `_walk_system` etc.).

File: src/aiperf/plugin/plugins.yaml [hunk 7]
```diff
@@ -237,6 +237,22 @@ endpoint:
       supports_audio: true
       metrics_title: LLM Metrics
 
+  messages:
```
Comment (ilana-n) [inline src/aiperf/plugin/plugins.yaml:239]: **Severity: Medium** Adding this new `messages` endpoint also needs a matching inference readiness probe entry. The existing probe table lives outside this diff in `src/aiperf/common/readiness_probe.py`, but this registration is what makes the gap observable: `_CANNED_PAYLOADS` and `_DEFAULT_PATHS` only cover `chat`, `completions`, and `embeddings`, so `--endpoint-type messages --wait-for-model-mode inference` falls back to posting the chat payload to `/v1/chat/completions`.

Repro:

```bash
uv run python -c "from aiperf.common.readiness_probe import _build_inference_probe_request; url, body = _build_inference_probe_request(url='http://mock.local', model_name='claude', endpoint_type='messages', custom_endpoint=None); print(url); print(body.decode())"
```

which prints:

```text
http://mock.local/v1/chat/completions
{"messages":[{"role":"user","content":"Lo"}],"max_tokens":1,"model":"claude"}
```

That can mark the wrong route as ready, especially because `_wait_inference` treats any status below 500 as ready, so a native Messages server returning 404/400 on `/v1/chat/completions` would still pass preflight before `/v1/messages` is actually ready. Could this PR add a `messages` canned payload and default path for `/v1/messages`?

**Comment by hnt2601** [comment]
Thanks @ajcasagrande for unifying this and preserving the authorship 🙏 Happy to keep contributing.

I can pick up the remaining open thread — @ilana-n's finding that the new `messages` endpoint has no matching inference readiness-probe entry. Confirmed on this branch: `_CANNED_PAYLOADS` and `_DEFAULT_PATHS` in `src/aiperf/common/readiness_probe.py` only cover `chat`/`completions`/`embeddings`, so `--endpoint-type messages --wait-for-model-mode inference` falls through to the chat-compatible fallback (posts a `/v1/chat/completions`-shaped body to a Messages-only server).

I have a small fix ready (TDD, red→green): add a dedicated `messages` template (`{messages, max_tokens: 1}`) + `/v1/messages` path, with a unit test asserting no chat fallback. Happy to open a PR into `ajc/anthropic-messages` — unless @ilana-n is already on it. Just say the word and I'll push.

**Comment by ajcasagrande** [comment]
@hnt2601 yes, that would be great, thanks!

**Comment by hnt2601** [comment]
Opened #1142 into `ajc/anthropic-messages` with the fix + test. @ilana-n this should resolve your readiness-probe thread — the probe now hits `/v1/messages` with a Messages-shaped body instead of the chat fallback.

**Comment by ajcasagrande** [comment]
@ilana-n ready for re-review after merging @hnt2601 's PR

File: src/aiperf/cli_runner/_preflight.py [hunk 8]
```diff
@@ -17,7 +17,25 @@
...
+_ANTHROPIC_VERSION = "2023-06-01"
+
+
+def _readiness_auth_headers(cfg: EndpointConfig) -> dict[str, str]:
+    """Build readiness-probe headers matching the endpoint's auth scheme."""
+    headers = dict(cfg.headers or {})
+    if str(cfg.type) == "messages":
+        headers.setdefault("anthropic-version", _ANTHROPIC_VERSION)
+        if cfg.api_key:
+            headers.setdefault("x-api-key", cfg.api_key)
```
Comment (ilana-n) [inline src/aiperf/cli_runner/_preflight.py]: Could we make this match `MessagesEndpoint.get_endpoint_headers()` exactly by assigning `headers["x-api-key"] = cfg.api_key` here instead of `setdefault`? The endpoint request path lets `--api-key` override any preconfigured `x-api-key` header, but readiness currently keeps the header value from `cfg.headers`. That creates an edge case where preflight probes one key while the actual Messages requests use another.

File: tests/aiperf_mock_server/app.py [hunk 9]
```diff
@@ -379,6 +383,186 @@ async def _chat_stream_wrapper(
...
+        "usage": {
+            "input_tokens": input_tokens,
+            "output_tokens": ctx.usage["completion_tokens"],
+            "cache_creation_input_tokens": cache_creation,
+            "cache_read_input_tokens": cache_read,
+        },
+    }
+
+
+@app.post("/v1/messages", response_model=None)
```
Comment (ilana-n) [inline tests/aiperf_mock_server/app.py:512]: Since this adds `/v1/messages` as a mock inference route, should we also add that path to `_INFERENCE_PATHS` and `_AUTH_PROTECTED_PATHS` above? Otherwise the new Messages route does not get the mock readiness-delay behavior, and a mock server started with an API key will still allow `/v1/messages` requests through as long as `anthropic-version` is present.

File: docs/tutorials/anthropic-messages-endpoint.md [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+
+## Basic Usage
+
+### Non-Streaming
+
+```bash
+aiperf profile \
+    --model claude-sonnet-4-20250514 \
+    --endpoint-type messages \
+    --url api.anthropic.com \
```
Comment (ilana-n) [inline docs/tutorials/anthropic-messages-endpoint.md]: Could these examples use the official Claude API base URL, `https://api.anthropic.com`, instead of the schemeless `api.anthropic.com`? That makes the copy/paste examples line up with Anthropic’s docs and avoids depending on AIPerf’s URL normalization behavior.

**Review by ilana-n — APPROVED** [review]
small comments added, overall LGTM!
